crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -55,12 +69,53 @@ impl ArrayHeap {
         self.staged.clear();
     }
 
-    fn push(&mut self, index: Expression, value: Expression, op: Expression) {
-        let item = MemOp { operation: op, value, index };
+    fn push(&mut self, item: MemOp) {
+        let is_load = item.operation == Expression::zero();
+        let index_const = item.index.to_const();
+        self.typ = match &self.typ {
+            ArrayType::Init(init_idx, len) => match (is_load, index_const) {
+                (false, Some(idx)) => {
+                    let idx: MemAddress = idx.to_u128().try_into().unwrap();
+                    let mut init_idx2 = init_idx.clone();
+                    let mut len2 = *len;
+                    init_idx2.insert(idx);
+                    if idx + 1 > len2 {
+                        len2 = idx + 1;
+                    }
+                    ArrayType::Init(init_idx2, len2)
+                }
+                (false, None) => ArrayType::WriteOnly,
+                (true, _) => {
+                    if *len as usize == init_idx.len() {
+                        ArrayType::ReadOnly(None)
+                    } else {
+                        ArrayType::ReadWrite(None)
+                    }
+                }
+            },
+            ArrayType::WriteOnly => {
+                if is_load {
+                    ArrayType::ReadWrite(None)
+                } else {
+                    ArrayType::WriteOnly
+                }
+            }
+            ArrayType::ReadOnly(last) => match (is_load, last) {
+                (true, Some(_)) => ArrayType::ReadWrite(None),
```

Comment:
Why does a previously read only array become read/write if its 'last' is Some?

Reply:
Last indicates the first write to the array (after the initialisation phase), it seems strange to be read-only with a write, but if we have only writes at the end, we can discard them as nobody will read them.

Reply:
> Last indicates the first write to the array

Why is it named `last` then?

Your explanation on the rest makes sense.

Reply:
It represent the position of the 'last writes'; it is the minimum position of the last writes